plugin/storage/cassandra/spanstore/reader.go

```diff
@@ -177,26 +238,135 @@ func (s *SpanReader) FindTraces(queryParams *spanstore.TraceQueryParameters) ([]
+	if traceQuery.DurationMin != 0 {
+		return s.queryByDuration(traceQuery)
+	}
+
+	// NB(rooz): consider refactoring the query... methods where they take in traceQuery instead of a breakdown of its values
+	if traceQuery.OperationName != "" {
+		traceIds, err := s.queryByServiceNameAndOperation(traceQuery.ServiceName, traceQuery.OperationName, traceQuery.StartTimeMin, traceQuery.StartTimeMax, traceQuery.NumTraces)
+		if err != nil {
+			return nil, err
+		}
+		if len(traceQuery.Tags) > 0 {
+			tagTraceIds, err := s.queryByTagsAndLogs(traceQuery.ServiceName, traceQuery.StartTimeMin, traceQuery.StartTimeMax, traceQuery.Tags, traceQuery.NumTraces)
+			if err != nil {
+				return nil, err
+			}
+			return s.intersectTraceIDSlices(traceIds, tagTraceIds), nil
+		}
+		return traceIds, err
+	}
+	if len(traceQuery.Tags) > 0 {
+		return s.queryByTagsAndLogs(traceQuery.ServiceName, traceQuery.StartTimeMin, traceQuery.StartTimeMax, traceQuery.Tags, traceQuery.NumTraces)
+	}
+	return s.queryByService(traceQuery.ServiceName, traceQuery.StartTimeMin, traceQuery.StartTimeMax, traceQuery.NumTraces)
+}
+
+func (s *SpanReader) intersectTraceIDSlices(one, other []dbmodel.TraceID) []dbmodel.TraceID {
+	oneAsMap := dbmodel.UniqueTraceIDs{}
+	for _, i := range one {
+		oneAsMap[i] = struct{}{}
+	}
+	var retMe []dbmodel.TraceID
+	for _, i := range other {
+		if _, ok := oneAsMap[i]; ok {
+			retMe = append(retMe, i)
+		}
+	}
+	return retMe
+}
+
+func (s *SpanReader) queryByTagsAndLogs(serviceName string, startTime, endTime time.Time, tags map[string]string, limit int) ([]dbmodel.TraceID, error) {
+	results := dbmodel.UniqueTraceIDs{}
+	for k, v := range tags {
+		// service_name = ? AND tag_key = ? AND tag_value = ? and start_time > ? and start_time < ? limit ?`
+		query := s.session.Query(
+			queryByTag,
+			serviceName,
+			k,
+			v,
+			model.TimeAsEpochMicroseconds(startTime),
+			model.TimeAsEpochMicroseconds(endTime),
+			limit,
+		)
+		t, err := s.executeQuery(query, s.readerMetrics.queryTagIndex)
+		if err != nil {
+			return nil, err
+		}
+		for _, tID := range t {
+			results.Add(tID)
+		}
+	}
+	return results.ToList(), nil
+}
+
+func (s *SpanReader) queryByDuration(traceQuery *spanstore.TraceQueryParameters) ([]dbmodel.TraceID, error) {
+	results := dbmodel.UniqueTraceIDs{}
+
+	minDurationMicros := (traceQuery.DurationMin.Nanoseconds() / int64(time.Microsecond/time.Nanosecond))
+	maxDurationMicros := (traceQuery.DurationMax.Nanoseconds() / int64(time.Microsecond/time.Nanosecond))
+
+	// See writer.go:indexByDuration  for how this is indexed
+	// This is indexed in hours since epoch, converted to seconds
+	// TODO encapsulate this calculation
+	startTimeSeconds := (traceQuery.StartTimeMin.UnixNano() / int64(time.Hour))
+	endTimeSeconds := (traceQuery.StartTimeMax.UnixNano() / int64(time.Hour))
+
+	for timeBucket := endTimeSeconds; timeBucket >= startTimeSeconds; timeBucket-- {
```

Comment:
after looking more at the writer, this logic does seem correct, just the variable names are messed up (startTimeSeconds is actually startTimeHour). But we should redo it as normal timestamps per #81